imap/reconstruct.c

```diff
@@ -715,3 +732,151 @@ static void do_mboxlist(void)
+    }
+    header_path = buf_cstring(&buf);
+
+    /* Create an mbentry from cyrus.header */
+    mbentry = mailbox_mbentry_from_path(header_path);
+    if (!mbentry) {
+        printf("Invalid/missing cyrus.header at '%s'\n", header_path);
+        goto done;
+    }
+    else if (!mbentry->partition) {
+        printf("Header for mailbox '%s' is not located on a valid [meta]partition\n",
+               mbentry->name);
+        goto done;
+    }
+
+    namespacelock = mboxname_usernamespacelock(mbentry->name);
+
+    if (!mbentry->uniqueid) {
+        /* Look elsewhere for a UID */
+        const char *root;
+        if (config_metapartition_files & IMAP_ENUM_METAPARTITION_FILES_HEADER) {
+            root = config_metapartitiondir(mbentry->partition);
+        }
+        else {
+            root = config_partitiondir(mbentry->partition);
+        }
+        const char *p = header_path + strlen(root);
+        if (*p == '/') p++;
+
+        if (!strncmp(p, "uuid/", 5)) {
+            /* Try to gleen UID from path */
+            tok_t tok = TOK_INITIALIZER(p+5, "/", 0);
+            char *token = tok_next(&tok);
+
+            if (token && strlen(token) == 1) {
+                const char first = *token;
+
+                token = tok_next(&tok);
+                if (token && strlen(token) == 1) {
+                    const char second = *token;
+
+                    token = tok_next(&tok);
+                    if (token && strlen(token) >= 2 &&
+                        first == token[0] && second == token[1]) {
+                        mbentry->uniqueid = xstrdup(token);
+                        mbentry->mbtype &= ~(MBTYPE_LEGACY_DIRS);
+
+                        token = tok_next(&tok);
+                        if (strcmpnull(token, FNAME_HEADER+1) || tok_next(&tok)) {
+                            xzfree(mbentry->uniqueid);
+                        }
+                    }
+                }
+            }
+
+            tok_fini(&tok);
+
+            if (!mbentry->uniqueid) {
+                printf("Header for mailbox '%s' is not located in proper UUID location: '%s'\n",
+                       mbentry->name, header_path);
+                goto done;
+            }
+        }
+        else {
+            /* Look for an existing N record */
+            mbentry_t *mbentry_byname = NULL;
+
+            mboxlist_lookup_allow_all(mbentry->name, &mbentry_byname, NULL);
+            if (mbentry_byname) {
+                mbentry->uniqueid = xstrdupnull(mbentry_byname->uniqueid);
+            }
+            mboxlist_entry_free(&mbentry_byname);
+
+            if (!mbentry->uniqueid) {
+                /* Generate a new UID */
+                mbentry->uniqueid = xstrdup(makeuuid());
+            }
+
+            mbentry->mbtype |= MBTYPE_LEGACY_DIRS;
+        }
+    }
+
+    /* Sanity check the path */
+    int legacy = (mbentry->mbtype & MBTYPE_LEGACY_DIRS) ||
+        config_getswitch(IMAPOPT_MAILBOX_LEGACY_DIRS);
```

Comment:
I think the IMAPOPT_MAILBOX_LEGACY_DIRS check can be skipped here?

AIUI, that option determines whether _new_ mailboxes are legacy or uuid.  But in the `reconstruct -P` case we know we already have a mailbox on disk cause we have the path to the header file, we're just trying to recreate/repair an mbentry for it.  So all we care about is where it is now, not where it would be created if it were new.

The mbentry->mbtype will already have its MBTYPE_LEGACY_DIRS flag set or not-set correctly from the path examination above.

If the mailbox being reconstructed is already in a uuid path, but the system is set up for new mailboxes to be on a legacy path, then I think this check as is will fall through the `||`, and then accidentally set legacy to 1 because of the IMAPOPT_MAILBOX_LEGACY_DIRS, even though the path says this is a uuid-mailbox.  And then the metapath will be calculated wrong, and the "header ...  is not located in proper location" check will bail out.

Reply:
Good point.  Fixed